apollo-router/src/extensions/header_filter.rs

```diff
@@ -0,0 +1,68 @@
+use apollo_router_core::prelude::*;
+use std::{
+    collections::HashSet,
+    task::{Context, Poll},
+};
+
+use tower::{Layer, Service};
+
+pub struct HeaderFilterLayer {
+    allowed_headers: HashSet<String>,
+}
+
+impl HeaderFilterLayer {
+    pub fn new<I>(allowed_headers: I) -> Self
+    where
+        I: Iterator<Item = String>,
+    {
+        HeaderFilterLayer {
+            allowed_headers: allowed_headers.map(|s| s.to_lowercase()).collect(),
+        }
+    }
+}
+
+impl<S> Layer<S> for HeaderFilterLayer {
+    type Service = HeaderFilter<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        HeaderFilter {
+            service: inner,
+            allowed_headers: self.allowed_headers.clone(),
+        }
+    }
+}
+
+pub struct HeaderFilter<S> {
+    service: S,
+    allowed_headers: HashSet<String>,
+}
+
+impl<S> Service<graphql::RouterRequest> for HeaderFilter<S>
+where
+    S: Service<graphql::RouterRequest, Error = ()>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: graphql::RouterRequest) -> Self::Future {
+        let removed_keys: Vec<_> = req
+            .frontend_request
+            .headers()
+            .keys()
+            .filter(|name| !self.allowed_headers.contains(name.as_str()))
+            .map(|name| name.clone())
+            .collect();
+
+        let headers_mut = req.frontend_request.headers_mut();
+        for name in removed_keys {
+            headers_mut.remove(name);
+        }
+
+        self.service.call(req)
```

Comment:
Not what is the use of this extension :thinking: this would filter headers on the incoming request. We don't forward headers to the subgraph yet, do we plan to?

Reply:
we will need at some point, for authorization headers. With this extension filtering the client request, and subgraph extensions receiving the type defined in https://github.com/apollographql/router/issues/305#issuecomment-1015449844, they would be able to extract data from the request. We might not even need that filter, it is just an easy example to test the design